## Final Project Submission

Please fill out:
* Student name: Richard Gachiri Muriithi
* Student pace: part time 
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:


## Overview
This project analyzes the ......

## Data Understanding




**Loading data**

We load data into a data structure called a **dataframe**. A dataframe contains rows and columns; it can be easily manipulated hence appropriate for data analysis.

In [76]:
# Your code here - remember to use markdown cells for comments as well!
import numpy as np
import pandas as pd

In [77]:
df = pd.read_csv('./Data/imdb.title.ratings.csv.gz')

In [78]:
df.info() # checks for the overview of the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73856 entries, 0 to 73855
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   tconst         73856 non-null  object 
 1   averagerating  73856 non-null  float64
 2   numvotes       73856 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 1.7+ MB


In [79]:
df.shape # checks for the number of rows and columns

(73856, 3)

In [80]:
df.dtypes # dtype attribute checks the typer of data 

tconst            object
averagerating    float64
numvotes           int64
dtype: object

In [81]:
df.describe() # check for a statistical summary of the data

averagerating      numvotes
count   73856.000000  7.385600e+04
mean        6.332729  3.523662e+03
std         1.474978  3.029402e+04
min         1.000000  5.000000e+00
25%         5.500000  1.400000e+01
50%         6.500000  4.900000e+01
75%         7.400000  2.820000e+02
max        10.000000  1.841066e+06

In [82]:
df1 = pd.read_csv('./Data/imdb.title.basics.csv.gz')

In [83]:
df1.info() # checks for the overview of the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   tconst           146144 non-null  object 
 1   primary_title    146144 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres           140736 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 6.7+ MB


In [84]:
df1.shape # checks for the number of rows and columns

(146144, 6)

In [85]:
df1.dtypes # dtype attribute checks the typer of data 

tconst              object
primary_title       object
original_title      object
start_year           int64
runtime_minutes    float64
genres              object
dtype: object

In [86]:
df1.describe() # check for a statistical summary of the data

start_year  runtime_minutes
count  146144.000000    114405.000000
mean     2014.621798        86.187247
std         2.733583       166.360590
min      2010.000000         1.000000
25%      2012.000000        70.000000
50%      2015.000000        87.000000
75%      2017.000000        99.000000
max      2115.000000     51420.000000

In [87]:
df2 = pd.read_csv('./Data/bom.movie_gross.csv.gz')

In [88]:
df2.info() # checks for the overview of the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


In [89]:
df2.shape # checks for the number of rows and columns

(3387, 5)

In [90]:
df2.dtypes # dtype attribute checks the typer of data 

title              object
studio             object
domestic_gross    float64
foreign_gross      object
year                int64
dtype: object

In [91]:
df2.describe() # check for a statistical summary of the data

domestic_gross         year
count    3.359000e+03  3387.000000
mean     2.874585e+07  2013.958075
std      6.698250e+07     2.478141
min      1.000000e+02  2010.000000
25%      1.200000e+05  2012.000000
50%      1.400000e+06  2014.000000
75%      2.790000e+07  2016.000000
max      9.367000e+08  2018.000000

### The IMDB ratings, title and movie gross data

In [92]:
df.head()#checks for the first 5 rows

tconst  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [93]:
df.tconst.nunique() ## all the tconst are unique from each other

73856

In [94]:
df.averagerating.describe()

count    73856.000000
mean         6.332729
std          1.474978
min          1.000000
25%          5.500000
50%          6.500000
75%          7.400000
max         10.000000
Name: averagerating, dtype: float64

In [158]:
df.averagerating.value_counts()

7.0     2262
6.6     2251
7.2     2249
6.8     2239
6.5     2221
        ... 
9.6       18
10.0      16
9.8       15
9.7       12
9.9        5
Name: averagerating, Length: 91, dtype: int64

The df1 dataset, ratings contains movies with start years from 2010 to 2115 and also includes a wide varierty of primary title, original title and genres 

In [96]:
df1.head() #checks for the first 5 rows

tconst                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

In [97]:
df1.tconst.nunique() # the tconst are unique from each other

146144

In [98]:
df1.primary_title.value_counts()

Home                                                24
The Return                                          20
Broken                                              20
Alone                                               16
Homecoming                                          16
                                                    ..
Chidiya                                              1
Necessary Evil: Super-Villains of DC Comics          1
Karol's Message                                      1
I Skogen                                             1
Seeing From Within: The Life of Barbara Blackman     1
Name: primary_title, Length: 136071, dtype: int64

In [99]:
df1.original_title.value_counts()

Broken                                              19
Home                                                18
The Return                                          17
Freedom                                             13
Homecoming                                          13
                                                    ..
La Chasse au Collet                                  1
Death Fall                                           1
You'll Never Walk Alone                              1
70 80 90: A Shenzhen Love Story                      1
Seeing From Within: The Life of Barbara Blackman     1
Name: original_title, Length: 137773, dtype: int64

In [100]:
df1.start_year.value_counts()

2017    17504
2016    17272
2018    16849
2015    16243
2014    15589
2013    14709
2012    13787
2011    12900
2010    11849
2019     8379
2020      937
2021       83
2022       32
2023        5
2024        2
2027        1
2026        1
2025        1
2115        1
Name: start_year, dtype: int64

In [101]:
df1.genres.value_counts()[:10]

Documentary              32185
Drama                    21486
Comedy                    9177
Horror                    4372
Comedy,Drama              3519
Thriller                  3046
Action                    2219
Biography,Documentary     2115
Drama,Romance             2079
Comedy,Drama,Romance      1558
Name: genres, dtype: int64

The df2 dataset, Movie gross contains title movies from between the year 2010 to the year 2018. It also contains a wide variety of the movie titles, the studios, the domestics gross and foreign gross


In [102]:
df2.head() #checks for the first 5 rows

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [103]:
#df2['Years'] =  df2['year']
#df2['Years'].value_counts()

df2.year.value_counts()

2015    450
2016    436
2012    400
2011    399
2014    395
2013    350
2010    328
2017    321
2018    308
Name: year, dtype: int64

In [104]:
df2.title.value_counts()

Bluebeard                        2
To Die Like a Man                1
Stories We Tell                  1
Chicken with Plums               1
Max Steel                        1
                                ..
Katti Batti                      1
Brothers: Blood Against Blood    1
Tomboy (2011)                    1
Ready                            1
Teri Meri Kahaani                1
Name: title, Length: 3386, dtype: int64

All the title of the movies appear once except from the Bluebeard movie title

In [105]:
df2.studio.value_counts()

IFC             166
Uni.            147
WB              140
Magn.           136
Fox             136
               ... 
Conglomerate      1
Cleopatra         1
Proud             1
FEF               1
App.              1
Name: studio, Length: 257, dtype: int64

In [106]:
df2.domestic_gross.value_counts()

1100000.0      32
1000000.0      30
1300000.0      30
1200000.0      25
1400000.0      23
               ..
68800.0         1
87000000.0      1
739000.0        1
336000000.0     1
727000.0        1
Name: domestic_gross, Length: 1797, dtype: int64

In [107]:
df2.domestic_gross.describe()

count    3.359000e+03
mean     2.874585e+07
std      6.698250e+07
min      1.000000e+02
25%      1.200000e+05
50%      1.400000e+06
75%      2.790000e+07
max      9.367000e+08
Name: domestic_gross, dtype: float64

In [108]:
df2.foreign_gross.value_counts()

1200000      23
1100000      14
4200000      12
1900000      12
2500000      11
             ..
19800000      1
80100000      1
170600000     1
110000000     1
330000000     1
Name: foreign_gross, Length: 1204, dtype: int64

In [109]:
df2.foreign_gross.describe()

count        2037
unique       1204
top       1200000
freq           23
Name: foreign_gross, dtype: object

## Data Preparation

###  Data Cleaning
From the above data sets we will check first for missing data values in all the datasets. After identifying the missing values we will clean the data sets by either: 
1. Drop missing values, either columns or rows.
2. Fill/replace with mean/median/mode/, backward fill/forward fill, certain value
3. Interpolate

In [110]:
df.isnull().sum() # Gets count of missing values, imdb.title.ratings.csv

tconst           0
averagerating    0
numvotes         0
dtype: int64

We can see that from the dataset, imdb.title.ratings they is no missing values.

In [111]:
df1.isnull().sum() # Gets count of missing values, imdb.title.basics.csv

tconst                 0
primary_title          0
original_title        21
start_year             0
runtime_minutes    31739
genres              5408
dtype: int64

We can see that from the dataset, imdb.title.basics they is  missing values in:
1. original_title, 21 missing values
2. runtime_minutes, 31739 missing values
3. genres, 5408 missing values. 


In [112]:
df2.isnull().sum() # Gets count of missing values, born.movie_gross.cvs

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

We can see that from the dataset, born.movie_gross.cvs they is  missing values in:
1. studio, 5 missing values
2. domestic_gross, 28 missing values
3. foreign_gross, 1350 missing values. 

In [113]:
#imdb.title.ratings.csv
df.isna().sum() # Gets count of missing values

tconst           0
averagerating    0
numvotes         0
dtype: int64

In [114]:
#imdb.title.basics.csv
df1.isna().sum()# Gets count of missing values

tconst                 0
primary_title          0
original_title        21
start_year             0
runtime_minutes    31739
genres              5408
dtype: int64

In [115]:
#born.movie_gross.cvs
df2.isna().sum() # Gets count of missing values

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

In [116]:
#imdb.title.ratings.csv
missing_values = df.isna() # Checks missing values 
missing_values.head()

tconst  averagerating  numvotes
0   False          False     False
1   False          False     False
2   False          False     False
3   False          False     False
4   False          False     False

In [117]:
#imdb.title.basics.csv
missing_values = df1.isna() # Checks missing values
missing_values.head()

tconst  primary_title  original_title  start_year  runtime_minutes  genres
0   False          False           False       False            False   False
1   False          False           False       False            False   False
2   False          False           False       False            False   False
3   False          False           False       False             True   False
4   False          False           False       False            False   False

In [118]:
#born.movie_gross.cvs
missing_values = df2.isna() # Checks missing values
missing_values.head()

title  studio  domestic_gross  foreign_gross   year
0  False   False           False          False  False
1  False   False           False          False  False
2  False   False           False          False  False
3  False   False           False          False  False
4  False   False           False          False  False

Below we will check the percentage of missing value of the datasets **imdb.title.ratings, imdb.title.basic and born.movie_gross**. 

This will help use in knowing how to deal with the missing values in the datasets provided. 

In [119]:
def missing_values(data):
    # identify the total missing values per column
    # sort in order
    miss = data.isnull().sum().sort_values(ascending = False)

    # calculate percentage of the missing values
    percentage_miss = (data.isnull().sum() / len(data)).sort_values(ascending = False)

    # store in a dataframe
    missing = pd.DataFrame({"Missing Values": miss, "Percentage": percentage_miss}).reset_index()

    # remove values that are missing
    missing.drop(missing[missing["Percentage"] == 0].index, inplace = True)

    return missing

In [120]:
missing_data = missing_values(df)
missing_data

Empty DataFrame
Columns: [index, Missing Values, Percentage]
Index: []

From the df dataframe, imdb.title.ratings, they are no missing values at all.



In [121]:
missing_data1 = missing_values(df1)
missing_data1

index  Missing Values  Percentage
0  runtime_minutes           31739    0.217176
1           genres            5408    0.037005
2   original_title              21    0.000144

From the above results, df1 dataframe imdb.title.basics we can see the columns **rutime_minutes, genres, original title** have certain percentage of missing values. They are:
1. runtime_minutes 22%
2. genres 3.7%
3. original_title 0.0144%

In [122]:
missing_data2 = missing_values(df2)
missing_data2

index  Missing Values  Percentage
0   foreign_gross            1350    0.398583
1  domestic_gross              28    0.008267
2          studio               5    0.001476

From the above results, df2 dataframe bom.movie_gross we can see the columns **foreign_gross, domestic_gross, studio** have certain percentage of missing values. They are:
1. foreign_gross 39%
2. domestic_gross 0.8267%
3. studio 0.01476%

### Dealing with the missing values. 

#### df1 dataframe imdb.title.basics 


In [123]:
missing_data1

index  Missing Values  Percentage
0  runtime_minutes           31739    0.217176
1           genres            5408    0.037005
2   original_title              21    0.000144

In [125]:
df1.shape

(146144, 6)

For the column **runtime_mintues** with a very low percentage, we can replace it with the mean score of the runtime minutes

In [126]:
df1.runtime_minutes.fillna(df1.runtime_minutes.mean(), inplace = True)

since columns **genres, original title** have very low percentage we cand drop the rows where there is missing values

In [127]:
df1.dropna(axis = 0, subset=['genres'], inplace=True)
df1.dropna(axis = 0, subset=['original_title'], inplace=True)

## Using a function
#def drop_rows_missing_values(df, columns):
#  """Drops rows in columns with missing values.
#  simple function to drop the rows wtih missing values
#  """
#  return df.dropna(subset=columns, inplace = True)

#drop_rows = drop_rows_with_missing_values(df1,['genres','original_title'])

In [128]:
missing_values(df1)

Empty DataFrame
Columns: [index, Missing Values, Percentage]
Index: []

In [129]:
df1.isna().sum()

tconst             0
primary_title      0
original_title     0
start_year         0
runtime_minutes    0
genres             0
dtype: int64

In [130]:
df1.shape

(140734, 6)

#### df2 dataframe bom.movie_gross 

In [131]:
missing_data2

index  Missing Values  Percentage
0   foreign_gross            1350    0.398583
1  domestic_gross              28    0.008267
2          studio               5    0.001476

For the column **foreign_gross** with a very low percentage, we can replace it with the mean score of the runtime minutes

In [133]:
df2.foreign_gross

0       652000000
1       691300000
2       664300000
3       535700000
4       513900000
          ...    
3382          NaN
3383          NaN
3384          NaN
3385          NaN
3386          NaN
Name: foreign_gross, Length: 3387, dtype: object

In [134]:
df2.shape

(3387, 5)

For the column **domestic_gross, studio** with a we will drop the rows with the missing values

In [135]:
df2.dropna(axis = 0, subset=['domestic_gross'], inplace=True)
df2.dropna(axis = 0, subset=['studio'], inplace=True)


## Using a function to drop rows#
#def drop_rows_missing_values(df, columns):
#  """Drops rows in columns with missing values.
#  simple function to drop the rows wtih missing values
#  """
#  return df.dropna(axis = 0, subset=columns, inplace = True)

#drop_rows = drop_rows_with_missing_values(df2,['domestic_gross','studio'])

In [136]:
missing_values(df2)

index  Missing Values  Percentage
0  foreign_gross            1349    0.401967

For the column **foreign_gross** with a very low percentage, we can replace it with the median of its column. But the data type for this column is an object(string) thus we need to convert the datatype

In [137]:
df2.isna().sum()

title                0
studio               0
domestic_gross       0
foreign_gross     1349
year                 0
dtype: int64

Since for **foreign_gross** the data type is object, we covert it float64 so as to be able to replace the missing values using the median.

In [142]:
df2.foreign_gross = pd.to_numeric(df2.foreign_gross, errors= "coerce")

In [143]:
df2.dtypes

title              object
studio             object
domestic_gross    float64
foreign_gross     float64
year                int64
dtype: object

In [144]:
df2.foreign_gross.fillna(df2.foreign_gross.median(), inplace = True)

In [150]:
missing_values(df2)

Empty DataFrame
Columns: [index, Missing Values, Percentage]
Index: []

In [151]:
df2.isna().sum()

title             0
studio            0
domestic_gross    0
foreign_gross     0
year              0
dtype: int64

## Merging Datasets

In [152]:
merge_data = pd.merge(df, df1, on = 'tconst', how = 'inner')
merge_data

tconst  averagerating  numvotes  \
0      tt10356526            8.3        31   
1      tt10384606            8.9       559   
2       tt1042974            6.4        20   
3       tt1043726            4.2     50352   
4       tt1060240            6.5        21   
...           ...            ...       ...   
73047   tt9805820            8.1        25   
73048   tt9844256            7.5        24   
73049   tt9851050            4.7        14   
73050   tt9886934            7.0         5   
73051   tt9894098            6.3       128   

                                           primary_title  \
0                                       Laiye Je Yaarian   
1                                             Borderless   
2                                              Just Inès   
3                                 The Legend of Hercules   
4                                              Até Onde?   
...                                                  ...   
73047                                              Caisa   
73048  Code Geass: Lelouch of the Rebellion - Glorifi...   
73049                                            Sisters   
73050                                  The Projectionist   
73051                                             Sathru   

                                         original_title  start_year  \
0                                      Laiye Je Yaarian        2019   
1                                            Borderless        2019   
2                                             Just Inès        2010   
3                                The Legend of Hercules        2014   
4                                             Até Onde?        2011   
...                                                 ...         ...   
73047                                             Caisa        2018   
73048  Code Geass: Lelouch of the Rebellion Episode III        2018   
73049                                           Sisters        2019   
73050                                 The Projectionist        2019   
73051                                            Sathru        2019   

       runtime_minutes                    genres  
0           117.000000                   Romance  
1            87.000000               Documentary  
2            90.000000                     Drama  
3            99.000000  Action,Adventure,Fantasy  
4            73.000000          Mystery,Thriller  
...                ...                       ...  
73047        84.000000               Documentary  
73048       120.000000   Action,Animation,Sci-Fi  
73049        86.187247              Action,Drama  
73050        81.000000               Documentary  
73051       129.000000                  Thriller  

[73052 rows x 8 columns]

In [153]:
df2.rename(columns={'title': 'original_title'}, inplace=True)

In [156]:
merged_df3 = pd.merge(merge_data, df2, on= 'original_title', how = 'inner')
merged_df3

tconst  averagerating  numvotes  \
0     tt1043726            4.2     50352   
1     tt1171222            5.1      8296   
2     tt1210166            7.6    326657   
3     tt1212419            6.5     87288   
4     tt1229238            7.4    428142   
...         ...            ...       ...   
2419  tt3142688            5.8      5841   
2420  tt3399916            6.3      4185   
2421  tt3748512            7.4      4977   
2422  tt7008872            7.0     18768   
2423  tt8011712            7.4        54   

                             primary_title  \
0                   The Legend of Hercules   
1                            Baggage Claim   
2                                Moneyball   
3                                Hereafter   
4     Mission: Impossible - Ghost Protocol   
...                                    ...   
2419                         Finding Fanny   
2420                        The Dead Lands   
2421                    Hitchcock/Truffaut   
2422                            Boy Erased   
2423                              The Past   

                            original_title  start_year  runtime_minutes  \
0                   The Legend of Hercules        2014             99.0   
1                            Baggage Claim        2013             96.0   
2                                Moneyball        2011            133.0   
3                                Hereafter        2010            129.0   
4     Mission: Impossible - Ghost Protocol        2011            132.0   
...                                    ...         ...              ...   
2419                         Finding Fanny        2014            102.0   
2420                        The Dead Lands        2014            107.0   
2421                    Hitchcock/Truffaut        2015             79.0   
2422                            Boy Erased        2018            115.0   
2423                              The Past        2018            120.0   

                         genres studio  domestic_gross  foreign_gross  year  
0      Action,Adventure,Fantasy   LG/S      18800000.0     42400000.0  2014  
1                        Comedy   FoxS      21600000.0       887000.0  2013  
2         Biography,Drama,Sport   Sony      75600000.0     34600000.0  2011  
3         Drama,Fantasy,Romance     WB      32700000.0     72500000.0  2010  
4     Action,Adventure,Thriller   Par.     209400000.0    485300000.0  2011  
...                         ...    ...             ...            ...   ...  
2419     Adventure,Comedy,Drama    FIP        616000.0      7100000.0  2014  
2420           Action,Adventure  Magn.          5200.0     19600000.0  2015  
2421                Documentary  Cohen        260000.0     19600000.0  2015  
2422            Biography,Drama  Focus       6800000.0      5000000.0  2018  
2423               Drama,Horror    SPC       1300000.0      9300000.0  2013  

[2424 rows x 12 columns]

In [161]:
merged_df3.start_year.value_counts()

2014    317
2015    296
2011    295
2010    283
2012    280
2016    279
2013    257
2017    225
2018    181
2019     11
Name: start_year, dtype: int64

# Analysis